File: javascript/reverse_string.js
/*function reverseString(str) {
  const strsplit = str.split("")
  let newstr = ""
  strsplit.map(char =>{
    [char, newstr].join()
  })
  return newstr
}*/

function reverseString(str){
  let reversed = ""
  for (let i=str.length-1; i>-1; i--){
    reversed = reversed + str[i]
  }
  return reversed
}

/* EXPLANATION:
    i want to return every character in a string starting from the last character
    and decreasing the index until 0
    thus, instead of increasing i every time (i++), use i--
    and instead of i=0, start with i=str.length-1
    return the reversed string + character at [i]
*/

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'ih'");
  console.log("=>", reverseString("hi"));

  console.log("");

  console.log("Expecting: 'ybabtac'");
  console.log("=>", reverseString("catbaby"));
}

module.exports = reverseString;

// Please add your pseudocode to this file
// And a written explanation of your solution

//given str, divide that into individual characters
//map through the string and add each character to the beginning of a new string

File: ruby/reverse_string.rb
def reverse_string(str)
  # type your code in here
  reversed = ""
  str.chars.each do |char|
    reversed = char + reversed
  end
  reversed
end

# iterate over each char in str, and instead of doing 
# reverse + char, do char + reversed
# cat
# 

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'ih'"
  puts "=>", reverse_string('hi')

  puts

  puts "Expecting: 'ybabtac'"
  puts "=>", reverse_string('catbaby')

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution